File: src-tauri/src/commands/create_server.rs
use serde::Deserialize;
use std::{fs, path::Path};

use crate::{
    commands::download_server_jar::{download_server_jar, DownloadServerJarProps},
    utils::{create_folder::create_folder, run_server::run_server},
};

#[derive(Deserialize)]
pub struct CreateServerProps {
    name: String,
    description: String,
    version: String,
    server_dir: String,
    software: String,
    download_url: String,
}

#[tauri::command]
pub async fn create_server(props: CreateServerProps) -> Result<(), String> {
    let server_path = Path::new(&props.server_dir).join(&props.name);

    if server_path.exists() {
        return Err("Server already exists".to_string());
    }

    create_folder(server_path.to_str().unwrap()).await?;

    let server_props = serde_json::json!({
        "name": props.name,
        "description": props.description,
        "version": props.version,
        "ram": 1024,
        "software": props.software
    })
    .to_string();
    let server_props_path = server_path.join("satoru.json");
    fs::write(server_props_path, server_props).unwrap();

    download_server_jar(DownloadServerJarProps {
        server_dir: server_path.to_str().unwrap().to_string(),
        download_url: props.download_url,
    })
    .await?;

    let output = run_server(server_path.to_str().unwrap().to_string(), "1G".to_string());

    if let Ok(output) = output {
        if output.status.success() {
            Ok(())
        } else {
            Err(String::from_utf8_lossy(&output.stderr).to_string())
        }
    } else {
        Err("Failed to execute java command".to_string())
    }
}

File: src/lib/tauri/commands/create-server.ts
import { invoke } from "@tauri-apps/api";
import { ServerAlreadyExistsError } from "../errors/server-already-exists-error";

interface CreateServerProps {
  name: string;
  description: string;
  version: string;
  server_dir: string;
}

export async function createServer(props: CreateServerProps) {
  try {
    await invoke("create_server", { props });
  } catch (err) {
    if (typeof err !== "string") {
      throw new Error("Failed to create a server");
    }

    switch (err) {
      case "Server already exists":
        throw new ServerAlreadyExistsError();
      default:
        throw new Error(err);
    }
  }
}
